scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt

```diff
@@ -0,0 +1,208 @@
+     * identifier, as we try to lookup source code results if a GitHub repository is known.
+     */
+    private fun readFromClearlyDefined(
+        id: Identifier,
+        coordinates: ClearlyDefinedService.Coordinates
+    ): Result<ScanResultContainer> {
+        val startTime = Instant.now()
+        log.info("Looking up results for ${id.toCoordinates()}.")
+
+        return try {
+            val packageUrl = coordinates.toString()
+            val tools = execute(
+                clearlyDefinedService.harvestTools(
+                    coordinates.type,
+                    coordinates.provider,
+                    coordinates.namespace.orEmpty(),
+                    coordinates.name,
+                    coordinates.revision.orEmpty()
+                )
+            )
+            val scanCodeVersion = findScanCodeVersion(tools, packageUrl)
+            scanCodeVersion?.let { version ->
+                loadScanCodeResults(id, coordinates, version, startTime)
+            } ?: emptyResult(id)
+        } catch (e: IOException) {
+            log.error("Error when reading results for package $id from ClearlyDefined", e)
+            Failure("${e.javaClass}: ${e.message}")
+        }
+    }
+
+    /**
+     * Load the ScanCode results file for the package with the given [id] and [coordinates] from ClearlyDefined.
+     * The results have been produced by ScanCode in the given [version]; use the [startTime] for metadata.
+     */
+    private fun loadScanCodeResults(
+        id: Identifier,
+        coordinates: ClearlyDefinedService.Coordinates,
+        version: String,
+        startTime: Instant
+    ): Result<ScanResultContainer> {
+        val toolResponse = execute(
+            clearlyDefinedService.harvestToolData(
+                coordinates.type, coordinates.provider, coordinates.namespace.orEmpty(), coordinates.name,
+                coordinates.revision.orEmpty(), TOOL_SCAN_CODE, version
+            )
+        )
+        return toolResponse.use {
+            jsonMapper.readTree(it.byteStream())["content"]?.let { result ->
+                val summary = generateSummary(startTime, Instant.now(), "", result)
+                val details = generateScannerDetails(result)
+                Success(ScanResultContainer(id, listOf(ScanResult(Provenance(), details, summary))))
+            } ?: emptyResult(id)
+        }
+    }
+
+    /**
+     * Execute an HTTP request specified by the given [call]. The response status is checked. If everything went
+     * well, the marshalled body of the request is returned; otherwise, the function throws an exception.
+     */
+    private fun <T> execute(call: Call<T>): T {
+        val req = "${call.request().method} ${call.request().url}"
+        log.debug("Executing $req")
+        val response = call.execute()
+        log.debug("${response.code()} ${response.message()}")
+        if (response.isSuccessful) {
+            return response.body()!!
+        } else {
```

Comment:
The ` else` could be removed. If you should do that an empty line above log.error would be nice.